### PR TITLE
Add optional Logging and Monitoring to 3-GCVE stage

### PR DESCRIPTION
**Pull request by KonradSchieban**
---
**Checklist**

I applicable, I acknowledge that I have:
- [X] Read the [contributing guide](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/CONTRIBUTING.md)
- [X] Ran `terraform fmt` on all modified files
- [X] Regenerated the relevant README.md files using [`tools/tfdoc.py`](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/CONTRIBUTING.md#fabric-tools)
- [X] Made sure all relevant tests pass

### CODE REVIEW

File: modules/gcve-monitoring/main.tf [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+  base_gcve_agent_endpoint   = "https://storage.googleapis.com/gcve-observability-agent/latest/vmware-linux-amd64"
+  base_gcloud_secret_manager = "gcloud secrets versions access latest --secret="
+  sa_gcve_monitoring_roles = toset([
+    "roles/secretmanager.secretAccessor",
+    "roles/monitoring.admin",
+    "roles/logging.logWriter",
+  ])
+}
+
+data "google_compute_subnetwork" "gcve-subnetwork" {
```
Comment (juliocc) [inline modules/gcve-monitoring/main.tf:27]: We do not allow bare data sources, please gate it behind some sort of condition

File: modules/gcve-monitoring/main.tf [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+resource "google_project_iam_member" "gcve_monitoring_permissions" {
+  for_each = local.sa_gcve_monitoring_roles
+  role     = each.key
+  project  = var.project_id
+  member   = "serviceAccount:${google_service_account.sa_gcve_monitoring.email}"
+}
+
+resource "google_compute_firewall" "healthcheck" {
```
Comment (juliocc) [inline modules/gcve-monitoring/main.tf:45]: what if the user's environment already provides this?

File: modules/gcve-monitoring/main.tf [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  }
+
+  metadata = {
+    startup-script = templatefile("${path.module}/scripts/installer.sh",
+      {
+        endpoint_agent                 = "${local.base_gcve_agent_endpoint}/artifacts/bpagent-headless-vmware.tar.gz"
+        endpoint_install               = "${local.base_gcve_agent_endpoint}/installer/install.sh"
+        gcloud_secret_vsphere_server   = "${local.base_gcloud_secret_manager}${var.secret_vsphere_server}"
+        gcloud_secret_vsphere_user     = "${local.base_gcloud_secret_manager}${var.secret_vsphere_user}"
+        gcloud_secret_vsphere_password = "${local.base_gcloud_secret_manager}${var.secret_vsphere_password}"
```
Comment (juliocc) [inline modules/gcve-monitoring/main.tf:102]: metadata should not have sensitive data (it's visible from the UI)

**Review by juliocc — CHANGES_REQUESTED** [review]
@KonradSchieban thanks for your hard work on enhancing the GCVE FAST stage! I've taken a look and have some thoughts/suggestions below

The `gcv-monitoring` module you've created doesn't fit the typical structure of a Fabric module. Fabric modules are designed to be self-contained units that manage a single, specific resource (e.g., a VM, a VPC). In contrast, your component orchestrates the creation of multiple, diverse resources: projects, subnets, service accounts, secrets, etc. 

This broader scope indicates that it functions more like a blueprint. Blueprints in Fabric are intended for higher-level orchestration of multiple components or modules to achieve a complex deployment goal.

I suggest you refactor this PR to move the `gcve-monitoring` module into the existing `pc-minimal` blueprint, or maybe its own blueprint if you think that makes sense.

Please keep in mind these points:
- Avoid direct module-to-module calls within individual modules. (calling modules is OK for blueprints). Calling modules from within blueprints is acceptable (and often necessary)
- Allow users to pass references to existing resources (e.g., subnets, service accounts) as input parameters to your blueprint. If users don't provide these references, your blueprint should automatically create the required dependencies

File: modules/gcve-monitoring/README.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+Note that to complete the Monitoring setup you need to configure vCenter to send traffic to the Bindplane agent which listens by default on port TCP 5142.
+
+<!-- BEGIN TOC -->
+- [Basic Monitoring setup with default settings](#basic-monitoring-setup-with-default-settings)
+<!-- END TOC -->
+
+## Basic Monitoring setup with default settings
+
+```hcl
+module "gcve-monitoring" {
```
Comment (juliocc) [inline modules/gcve-monitoring/README.md:27]: this is not showing how to use the module

File: modules/gcve-monitoring/main.tf (hunk 3 above)
Comment (juliocc) [reply modules/gcve-monitoring/main.tf:102]: All good then :)

**Comment by KonradSchieban** [comment]
> @KonradSchieban thanks for your hard work on enhancing the GCVE FAST stage! I've taken a look and have some thoughts/suggestions below
> 
> The `gcv-monitoring` module you've created doesn't fit the typical structure of a Fabric module. Fabric modules are designed to be self-contained units that manage a single, specific resource (e.g., a VM, a VPC). In contrast, your module orchestrates the creation of multiple, diverse resources: projects, subnets, service accounts, secrets, etc.
> 
> This broader scope indicates that it functions more like a blueprint. Blueprints in Fabric are intended for higher-level orchestration of multiple components or modules to achieve a complex deployment goal.
> 
> I suggest you refactor this PR to move the `gcve-monitoring` module into the existing `pc-minimal` blueprint, or into its own blueprint if you think that makes sense.
> 
> Please keep in mind these points:
> 
> * Avoid direct module-to-module calls within individual modules. Calling modules from within blueprints is acceptable (and often necessary)
> * Allow users to pass references to existing resources (e.g., subnets, service accounts) as input parameters to your blueprint. If users don't provide these references, your blueprint should automatically create the required dependencies

Thanks for your review @juliocc ! It will take some time to make this changes but I'll work on it. I'll keep this PR open in the meantime and keep you updated 👍

**Comment by juliocc** [comment]
> Thanks for your review @juliocc ! It will take some time to make this changes but I'll work on it. I'll keep this PR open in the meantime and keep you updated 👍

Absolutely. Ping us if you need any help.

I'll close this in the meantime, but feel free to reopen once you're ready

**Pull request closed without merging**